### DISCUSSION TITLE
Cannot GraphQL delete and query for nested items

### DISCUSSION THREAD

**Discussion by ashih42** [prisma/prisma#4225] [Q&A] [answered]
Hello,

I am using the latest `@prisma/cli@2.11.0` and `@prisma/client@2.11.0` with `@nexus/schema@0.17.0`, `nexus-plugin-prisma@0.23.1`, `apollo-server@2.19.0`, but this is an old problem I've lived with for a long time throughout older versions.

I have the the following schema, where `Category` and `Transaction` have a many-to-many relationship stored in `CategoryEntry`:

```prisma
model Category {
  id  String  @id  @default(cuid())
  name  String
}

model Transaction {
  id  String  @id  @default(cuid())
  amount  Int
}

model CategoryEntry {
  id  String  @id  @default(cuid())

  category  Category  @relation(fields: [category_id], references: [id])
  category_id  String
  
  transaction  Transaction  @relation(fields: [transaction_id], references: [id])
  transaction_id  String

}
```

I have this GraphQL mutation to delete the specified `CategoryEntry`:
```ts
export const delete_category_entry = mutationField('delete_category_entry ',
  {
    type: 'CategoryEntry',
    args: {
      entry_id: stringArg({ required: true }),
    },
    resolve: async (parent, args, context) => {
      const { entry_id } = args;
      return await PRISMA_CLIENT.categoryEntry.delete({ where: { id: entry_id } });
    },
  },
);
```

Here's the problem when you call `delete_category_entry` and ask for specific return values.

**Example 1:** If I call `delete_category_entry` and request these return fields (only `id`), it is OK.
```graphql
mutation delete_category_entry($entry_id: String!) {
  delete_category_entry(entry_id: $entry_id) {
    id
  }
}
```

**Example 2:** However, If I call `delete_category_entry` and request these return fields (`id`, `category`, and `transaction`), it will delete the `CategoryEntry` successfully, but then it will give me this error when it tries to generate the return values.
```graphql
mutation delete_category_entry($entry_id: String!) {
  delete_category_entry(entry_id: $entry_id) {
    id
    category {
      name
    }
    transaction {
      amount
    }
  }
}
```

Error:
```
GraphQLError [Object]: Cannot return null for non-nullable field CategoryEntry.category.
```

I think this is happening because in this case, I deleted this `CategoryEntry`, and then I asked about this entry's `CategoryEntry.category` and `CategoryEntry.transaction` fields, so Prisma tried to do another database query using the `entry_id` value which has been deleted at this point, so it cannot find the associated `Category` and `Transaction` to fill the GraphQL return values.

However, in my `delete_category_entry` implementation, the object returned by `return await PRISMA_CLIENT.categoryEntry.delete({ where: { id: entry_id } });` does contain the `category_id` and `transaction_id`, so why can't Prisma just use these values to find the associated `Category` and `Transaction`?

Please suggest what we can do about this problem.  Thank you for your time and consideration.

**Comment by ryands17** [comment]
Hey @ashih42 👋 
In this case, you would need to create a custom resolver as Nexus automatically queries for your relations.

I personally achieved it in the following manner:

```prisma
model Category {
  id              String          @id @default(cuid())
  name            String
  categoryEntries CategoryEntry[]
}

model Transaction {
  id              String          @id @default(cuid())
  amount          Int
  categoryEntries CategoryEntry[]
}

model CategoryEntry {
  id String @id @default(cuid())

  category    Category @relation(fields: [category_id], references: [id])
  category_id String

  transaction    Transaction @relation(fields: [transaction_id], references: [id])
  transaction_id String
}
```

Using the above schema, I created the following models in Nexus:

```ts
import * as ns from '@nexus/schema'

const Category = ns.objectType({
  name: 'Category',
  definition(t) {
    t.model.id()
    t.model.name()
    t.model.categoryEntries()
  },
})

const Transaction = ns.objectType({
  name: 'Transaction',
  definition(t) {
    t.model.id()
    t.model.amount()
    t.model.categoryEntries()
  },
})

const CategoryEntry = ns.objectType({
  name: 'CategoryEntry',
  definition(t) {
    t.model.id()
    t.model.category()
    t.model.transaction()
  },
})

  // custom model
const CategoryEntryResolve = ns.objectType({
  name: 'CategoryEntryResolve',
  definition(t) {
    t.id('id')
    t.field('category', { type: 'Category' })
    t.field('transaction', { type: 'Transaction' })
  },
})

const Mutation = ns.extendType({
  type: 'Mutation',
  definition(t) {
    t.crud.createOneCategory()
    t.crud.createOneTransaction()
  },
})

const deleteCategoryEntry = ns.mutationField('deleteCategoryEntry', {
  // custom model
  type: 'CategoryEntryResolve',
  args: {
    entry_id: ns.stringArg({ required: true }),
  },
  resolve: async (_parent, args, ctx) => {
    const { entry_id } = args
    const data = await ctx.prisma.categoryEntry.delete({
      where: { id: entry_id },
      include: { category: true, transaction: true },
    })
    return data
  },
})
```

Now when I run the following query:

```graphql
mutation {
  deleteCategoryEntry(entry_id: "ckhdg0hgv00288ipkxb4ryal1") {
    id
    category {
      name
    }
    transaction {
      amount
    }
  }
}
```

I get the following result:
![image](https://user-images.githubusercontent.com/19697099/98817884-d2f22400-2450-11eb-931a-ed713b5e8df1.png)

**Comment by ashih42** [comment, marked as answer]
@ryands17 Thank you for your suggestion.  I did not know about the `include` option.  I understand how it works now, but I do not like this coupling of explicitly anticipating what the user might ask for in a place where I just want to implement the deletion logic.

From your answer, I realized I could use this simpler solution.  All I need is to define a custom resolver.  I changed my `CategoryEntry` as shown below:

```
export const CategoryEntry = objectType({
  name: 'CategoryEntry',
  definition(t: any) {
    t.model.id();

    // t.model.category();

    t.field('category', {
      type: 'Category',
      async resolve(root: any, args: any, context: any, info: any) {
        return await PRISMA_CLIENT.category.findOne({
          where: { id: root.category_id },
        });
      },
    });

  },
});
```

I think this resolver returns exactly what `t.model.category()` does, with additional benefit of still working even when this CategoryEntry database record has been deleted (it only needs `category_id` value, which the `PRISMA_CLIENT.category.delete()` call provides by default).  I think `t.model.category()` involves looking up the `CategoryEntry` table, resulting in a miss, whereas this custom resolver only look up the `Category` table , so it still works, and it might be slightly better in performance too.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
